<a href="https://colab.research.google.com/github/tdiffendal/USAT/blob/master/Copy_of_census_responses.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Examine census response rates, with data at state and census tract levels
## 2020 response rates from: https://2020census.gov/en/response-rates.html
## 2010 response rates from: https://api.census.gov/data/2010/dec/responserate/variables.html
## Demographic information in 2014-2018 ACS 5-year-estimate from: https://data2.nhgis.org/main

In [4]:
import pandas as pd
import numpy as np

# in excel changed original dates from m/d/yyyy to yyyy-mm-dd
initial_df = pd.read_csv('https://raw.githubusercontent.com/tdiffendal/USAT/master/census-responses/decennialrr2020_working.csv')
# had to resave as UTF-8 CSV, hence the 2
crosswalk = pd.read_csv('https://raw.githubusercontent.com/tdiffendal/USAT/master/census-responses/decennialrr2020_crosswalkfile2.csv')
# states paired with region as defined by census map at https://www2.census.gov/geo/pdfs/maps-data/maps/reference/us_regdiv.pdf
regions = pd.read_csv('https://raw.githubusercontent.com/tdiffendal/USAT/master/census-responses/state_region.csv')

## Columns


GEO_ID = Geographic Identifier

RESP_DATE = Posting Date

State = name of state (one of the 50 states, District of Columbia, Puerto Rico, or NaN)

Geo_Name = name of the tract, county, state

Region = region of the U.S. in which state is located as defined by census map at https://www2.census.gov/geo/pdfs/maps-data/maps/reference/us_regdiv.pdf

Geo_Type = type of geography; possible answers include Census Tract, Congressional District, Consolidated City, Country, County, County Subdivision, Place, Region, State, Tribal Tract, Tribal Area

DRRINT = Daily Self-Response Rate - Internet

DRRALL = Daily Self-Response Rate – Overall

CRRINT = Cumulative Self-Response Rate - Internet

not_int = new calculated column showing response rate NOT from internet

CRRALL = Cumulative Self-Response Rate – Overall

DINTMIN = Minimum Daily Internet Self-Response Rate

DMIN = Minimum Daily Overall Self-Response Rate

CINTMIN = Minimum Cumulative Internet Self-Response Rate

CMIN = Minimum Cumulative Overall Self-Response Rate

DINTMAX = Maximum Daily Internet Self-Response Rate

DMAX = Maximum Daily Overall Self-Response Rate

CINTMAX = Maximum Cumulative Internet Self-Response Rate

CMAX = Maximum Cumulative Overall Self-Response Rate

DINTAVG = Average Daily Internet Self-Response Rate

DAVG = Average Daily Overall Self-Response Rate

CINTAVG = Average Cumulative Internet Self-Response Rate

CAVG = Average Cumulative Overall Self-Response Rate

DINTMED = Median Daily Internet Self-Response Rate

DMED = Median Daily Overall Self-Response Rate

CINTMED = Median Cumulative Internet Self-Response Rate

CMED = Median Cumulative Overall Self-Response Rate

In [5]:
# merge responses and crosswalk
merged1 = pd.merge(initial_df, crosswalk, on='GEO_ID')

#merge merged1 with region data
merged = pd.merge(merged1, regions, on='State')

# create column showing responses not from internet
merged['not_int'] = merged.CRRALL - merged.CRRINT

#reorder columns to move State, Geo_Name and Geo_Type to front; also going to drop some values
cols = merged.columns.tolist()
cols = ['GEO_ID', 'RESP_DATE', 'State', 'Geo_Name', 'Region', 'Geo_Type', 
        'CRRINT', 'not_int', 'CRRALL']
merged = merged[cols]
merged = merged.rename(columns={'CRRINT':'internet', 'CRRALL':'2020_rate'})
merged

GEO_ID  RESP_DATE  ... not_int 2020_rate
0                 0100000US  6/15/2020  ...    12.3      61.4
1                0200000US1  6/15/2020  ...    11.7      60.8
2                0200000US2  6/15/2020  ...    13.8      66.5
3                0200000US3  6/15/2020  ...    13.2      58.6
4                0200000US4  6/15/2020  ...     9.9      61.9
...                     ...        ...  ...     ...       ...
123245  2560000US4755T00100  6/15/2020  ...     3.1      35.8
123246  2560000US4760T00100  6/15/2020  ...     0.2      10.9
123247  2560000US4770T00100  6/15/2020  ...     0.4      17.8
123248  2560000US4785T00100  6/15/2020  ...     0.1       8.1
123249  2560000US4785T00200  6/15/2020  ...     0.1      10.3

[123250 rows x 9 columns]

## States

### Read in 2020 data

In [6]:
# create df with response rate by state
is_states = merged['Geo_Type'] == 'State'
states = merged[is_states]
states = states.rename(columns={"internet": "state_internet", "not_int" : "state_not_int", "2020_rate": "2020_state_rate"})

# print df and sort by highest cumulative response rate
states.sort_values(by='2020_state_rate', ascending=False)

GEO_ID  RESP_DATE  ... state_not_int 2020_state_rate
55595   0400000US27  6/15/2020  ...          11.2            70.9
117774  0400000US55  6/15/2020  ...          12.0            68.4
50831   0400000US26  6/15/2020  ...          14.7            67.7
38079   0400000US19  6/15/2020  ...          14.4            67.7
64533   0400000US31  6/15/2020  ...          14.0            67.6
114850  0400000US53  6/15/2020  ...           8.9            66.8
112324  0400000US51  6/15/2020  ...          12.0            66.3
28633   0400000US17  6/15/2020  ...          12.1            66.1
82562   0400000US39  6/15/2020  ...          15.1            66.1
34724   0400000US18  6/15/2020  ...          15.3            65.9
110847  0400000US49  6/15/2020  ...           5.8            65.7
47228   0400000US24  6/15/2020  ...           9.5            65.4
40003   0400000US20  6/15/2020  ...          14.1            65.3
42735   0400000US21  6/15/2020  ...          17.1            65.0
27931   0400000US16  6/15/2020  ...          10.7            64.9
16869   0400000US09  6/15/2020  ...          11.3            64.8
91181   0400000US42  6/15/2020  ...          14.2            64.7
15097   0400000US08  6/15/2020  ...           8.8            64.6
89907   0400000US41  6/15/2020  ...          10.8            63.9
48873   0400000US25  6/15/2020  ...          10.0            63.4
67437   0400000US34  6/15/2020  ...           9.6            63.3
5448    0400000US06  6/15/2020  ...          10.2            62.4
99143   0400000US46  6/15/2020  ...          13.5            62.4
61053   0400000US29  6/15/2020  ...          14.5            61.6
81089   0400000US38  6/15/2020  ...          11.9            61.3
100434  0400000US47  6/15/2020  ...          15.8            61.0
66032   0400000US32  6/15/2020  ...          10.1            60.7
66842   0400000US33  6/15/2020  ...           9.6            60.6
97218   0400000US44  6/15/2020  ...          12.4            59.3
17931   0400000US10  6/15/2020  ...          10.7            59.3
5       0400000US01  6/15/2020  ...          18.1            59.2
18456   0400000US12  6/15/2020  ...          12.7            58.6
2197    0400000US04  6/15/2020  ...          10.3            57.9
77793   0400000US37  6/15/2020  ...          12.6            57.7
24001   0400000US13  6/15/2020  ...          11.4            57.4
18250   0400000US11  6/15/2020  ...           7.5            57.4
27501   0400000US15  6/15/2020  ...          10.8            56.8
59796   0400000US28  6/15/2020  ...          22.4            56.6
70925   0400000US36  6/15/2020  ...          11.4            56.6
4047    0400000US05  6/15/2020  ...          18.9            56.3
102563  0400000US48  6/15/2020  ...          10.3            56.1
97521   0400000US45  6/15/2020  ...          14.5            55.9
88050   0400000US40  6/15/2020  ...          14.9            55.8
44576   0400000US22  6/15/2020  ...          15.8            55.7
64029   0400000US30  6/15/2020  ...          13.5            54.9
121227  0400000US56  6/15/2020  ...          11.6            54.5
111836  0400000US50  6/15/2020  ...          12.5            54.3
116938  0400000US54  6/15/2020  ...          18.2            53.0
46322   0400000US23  6/15/2020  ...          14.1            52.8
70196   0400000US35  6/15/2020  ...          11.1            49.8
1970    0400000US02  6/15/2020  ...           8.6            46.8
121508  0400000US72  6/15/2020  ...           8.4            18.9

[52 rows x 9 columns]

### Read in 2010 states and join

In [52]:
# read in csvs with 2010 response data for states
states2010 = pd.read_csv('https://raw.githubusercontent.com/tdiffendal/USAT/master/census-responses/states2010.csv')

# merge with 2020 states
states_merged = pd.merge(states, states2010, on='State')
#get the column names
cols = states_merged.columns.tolist()
#only select columns we want
cols = ['GEO_ID', 'State', 'Region',
 '2020_state_rate', '2010_rate', '2000_rate']
states_merged = states_merged[cols]
states_merged = states_merged.rename(columns={'2000_rate':'2000_state_rate', '2010_rate':'2010_state_rate'})

#create column with difference in 2010 vs 2020 response rate
states_merged['10_20_state_difference'] = states_merged['2020_state_rate'] - states_merged['2010_state_rate']

#print table sorted by 10-20 difference largest ---> smallest
states_merged.sort_values(by='10_20_state_difference', ascending=False)

GEO_ID                 State       Region  2020_state_rate  \
5   0400000US08              Colorado         West             64.6   
47  0400000US53            Washington         West             66.8   
44  0400000US49                  Utah         West             65.7   
18  0400000US22             Louisiana        South             55.7   
27  0400000US31              Nebraska      Midwest             67.6   
13  0400000US17              Illinois      Midwest             66.1   
23  0400000US27             Minnesota      Midwest             70.9   
22  0400000US26              Michigan      Midwest             67.7   
28  0400000US32                Nevada         West             60.7   
20  0400000US24              Maryland        South             65.4   
4   0400000US06            California         West             62.4   
30  0400000US34            New Jersey    Northeast             63.3   
16  0400000US20                Kansas      Midwest             65.3   
12  0400000US16                 Idaho         West             64.9   
2   0400000US04               Arizona         West             57.9   
6   0400000US09           Connecticut    Northeast             64.8   
11  0400000US15                Hawaii         West             56.8   
15  0400000US19                  Iowa      Midwest             67.7   
21  0400000US25         Massachusetts      Midwest             63.4   
46  0400000US51              Virginia        South             66.3   
35  0400000US39                  Ohio      Midwest             66.1   
17  0400000US21              Kentucky        South             65.0   
48  0400000US54         West Virginia        South             53.0   
37  0400000US41                Oregon         West             63.9   
36  0400000US40              Oklahoma        South             55.8   
38  0400000US42          Pennsylvania    Northeast             64.7   
24  0400000US28           Mississippi        South             56.6   
25  0400000US29              Missouri      Midwest             61.6   
32  0400000US36              New York    Northeast             56.6   
29  0400000US33         New Hampshire    Northeast             60.6   
34  0400000US38          North Dakota      Midwest             61.3   
7   0400000US10              Delaware        South             59.3   
3   0400000US05              Arkansas        South             56.3   
0   0400000US01               Alabama        South             59.2   
14  0400000US18               Indiana      Midwest             65.9   
26  0400000US30               Montana         West             54.9   
49  0400000US55             Wisconsin      Midwest             68.4   
41  0400000US46          South Dakota      Midwest             62.4   
39  0400000US44          Rhode Island    Northeast             59.3   
50  0400000US56               Wyoming         West             54.5   
10  0400000US13               Georgia        South             57.4   
8   0400000US11  District of Columbia        South             57.4   
45  0400000US50               Vermont    Northeast             54.3   
43  0400000US48                 Texas        South             56.1   
42  0400000US47             Tennessee        South             61.0   
31  0400000US35            New Mexico         West             49.8   
19  0400000US23                 Maine    Northeast             52.8   
9   0400000US12               Florida        South             58.6   
1   0400000US02                Alaska         West             46.8   
33  0400000US37        North Carolina        South             57.7   
40  0400000US45        South Carolina        South             55.9   
51  0400000US72           Puerto Rico  Puerto Rico             18.9   

    2010_state_rate  2000_state_rate  10_20_state_difference  
5                72               75                    -7.4  
47               76               75                    -9.2  
44               75               74                    

## Census Tracts

In [54]:
# select just census tract geo types
tracts = merged[merged['Geo_Type'].str.contains("Tract")]
#rename column
tracts = tracts.rename(columns={"2020_rate": "2020_tract_rate"})
# sort by highest cumulative response rate
tracts.sort_values(by='2020_tract_rate', ascending=False)

GEO_ID  RESP_DATE          State  \
26459   1400000US13215010606  6/15/2020        Georgia   
49682   1400000US25013812903  6/15/2020  Massachusetts   
53285   1400000US26099223801  6/15/2020       Michigan   
113068  1400000US51059492202  6/15/2020       Virginia   
54097   1400000US26139021605  6/15/2020       Michigan   
...                      ...        ...            ...   
3401    1400000US04017940008  6/15/2020        Arizona   
122937   2560000US1440T00100  6/15/2020    Tribal Land   
73127   1400000US36047005303  6/15/2020       New York   
7920    1400000US06037277400  6/15/2020     California   
18595   1400000US12005000700  6/15/2020        Florida   

                       Geo_Name       Region      Geo_Type  internet  not_int  \
26459    Tract 106.06, Muscogee        South  Census Tract       0.0     98.1   
49682    Tract 8129.03, Hampden      Midwest  Census Tract      86.2      6.9   
53285     Tract 2238.01, Macomb      Midwest  Census Tract      86.2      6.6   
113068   Tract 4922.02, Fairfax        South  Census Tract      89.1      3.3   
54097      Tract 216.05, Ottawa      Midwest  Census Tract      85.4      7.0   
...                         ...          ...           ...       ...      ...   
3401      Tract 9400.08, Navajo         West  Census Tract       0.0      0.0   
122937          T001, Havasupai  Tribal Land  Tribal Tract       0.0      0.0   
73127        Tract 53.03, Kings    Northeast  Census Tract       0.0      0.0   
7920    Tract 2774, Los Angeles         West  Census Tract       0.0      0.0   
18595              Tract 7, Bay        South  Census Tract       0.0      0.0   

        2020_tract_rate  
26459              98.1  
49682              93.1  
53285              92.8  
113068             92.4  
54097              92.4  
...                 ...  
3401                0.0  
122937              0.0  
73127               0.0  
7920                0.0  
18595               0.0  

[84519 rows x 9 columns]

In [55]:
#tract rates compared to state averages
tract_v_state = pd.merge(tracts, states_merged, on=['State', 'Region'])
tract_v_state = tract_v_state[['GEO_ID_x', 'State', 'Geo_Name', 'Region','2020_tract_rate', '2020_state_rate', '2010_state_rate', '10_20_state_difference']]
tract_v_state = tract_v_state.rename(columns={'GEO_ID_x':'GEO_ID'})
tract_v_state['2020_tract_st_diff'] = tract_v_state['2020_tract_rate'] - tract_v_state['2020_state_rate']
tract_v_state.sort_values(by=['2020_tract_st_diff'])
#merging with tracts with states will drop tribal tracts (as they have no state), so those are examined separately below

GEO_ID       State                 Geo_Name   Region  \
38744  1400000US26099982100    Michigan       Tract 9821, Macomb  Midwest   
38379  1400000US26083980100    Michigan     Tract 9801, Keweenaw  Midwest   
40422  1400000US27007940001   Minnesota  Tract 9400.01, Beltrami  Midwest   
76533  1400000US49037942100        Utah     Tract 9421, San Juan     West   
76532  1400000US49037942000        Utah     Tract 9420, San Juan     West   
...                     ...         ...                      ...      ...   
48992  1400000US35049010602  New Mexico   Tract 106.02, Santa Fe     West   
71352  1400000US48121021535       Texas     Tract 215.35, Denton    South   
48799  1400000US35028000500  New Mexico      Tract 5, Los Alamos     West   
48570  1400000US35001003741  New Mexico  Tract 37.41, Bernalillo     West   
23292  1400000US13215010606     Georgia   Tract 106.06, Muscogee    South   

       2020_tract_rate  2020_state_rate  2010_state_rate  \
38744              0.0             67.7               78   
38379              0.0             67.7               78   
40422              4.7             70.9               81   
76533              0.6             65.7               75   
76532              0.8             65.7               75   
...                ...              ...              ...   
48992             85.4             49.8               65   
71352             91.9             56.1               71   
48799             86.9             49.8               65   
48570             87.2             49.8               65   
23292             98.1             57.4               72   

       10_20_state_difference  2020_tract_st_diff  
38744                   -10.3               -67.7  
38379                   -10.3               -67.7  
40422                   -10.1               -66.2  
76533                    -9.3               -65.1  
76532                    -9.3               -64.9  
...                       ...                 ...  
48992                   -15.2                35.6  
71352                   -14.9                35.8  
48799                   -15.2                37.1  
48570                   -15.2                37.4  
23292                   -14.6                40.7  

[84093 rows x 9 columns]

### Read in 2010 tract rates and join

In [56]:
# read in csvs with 2010 response data for tracts and states
rates2010 = pd.read_csv('https://raw.githubusercontent.com/tdiffendal/USAT/master/census-responses/2010responserate.csv')
#rename this column
rates2010 = rates2010.rename(columns={'FSRR2010':'2010_rate'})
rates2010

NAME            county         State  \
0          Census Tract 201    Autauga County       Alabama   
1          Census Tract 202    Autauga County       Alabama   
2          Census Tract 203    Autauga County       Alabama   
3          Census Tract 204    Autauga County       Alabama   
4       Census Tract 205.01    Autauga County       Alabama   
...                     ...               ...           ...   
84514  Census Tract 7505.01   Yauco Municipio   Puerto Rico   
84515  Census Tract 7505.02   Yauco Municipio   Puerto Rico   
84516  Census Tract 7505.03   Yauco Municipio   Puerto Rico   
84517  Census Tract 7506.01   Yauco Municipio   Puerto Rico   
84518  Census Tract 7506.02   Yauco Municipio   Puerto Rico   

                     GEO_ID  2010_rate  state_num  county_num   tract  
0      1400000US01001020100       70.6          1           1   20100  
1      1400000US01001020200       70.1          1           1   20200  
2      1400000US01001020300       73.6          1           1   20300  
3      1400000US01001020400       78.4          1           1   20400  
4      1400000US01001020501       81.2          1           1   20501  
...                     ...        ...        ...         ...     ...  
84514  1400000US72153750501       70.4         72         153  750501  
84515  1400000US72153750502       73.4         72         153  750502  
84516  1400000US72153750503       63.4         72         153  750503  
84517  1400000US72153750601       67.1         72         153  750601  
84518  1400000US72153750602       57.8         72         153  750602  

[84519 rows x 8 columns]

In [57]:
## difference in row numbers: both tract dfs have 84519 rows, but when joined only 84093
# Identify what values are in rates2010 and not in tracts
key_diff1 = set(rates2010.GEO_ID).difference(tracts.GEO_ID)
len(key_diff1)
#key_diff1

# Identify what values are in tracts and not in rates2010
key_diff2 = set(tracts.GEO_ID).difference(rates2010.GEO_ID)
len(key_diff2)
#key_diff2

426

##### Cleaning Conclusion: 
2010 rates do not include tribal tracts while 2020 tracts are missing some tracts in a multitude of states, likely due to a change in tract boundaries. Those differences account for 426 tracts, which is .5% of the original 84519 tracts. As these tracts are small percentage of all tracts, they can be dropped. 

In [58]:
# merge with 2020 tracts
tracts_merged = pd.merge(tract_v_state, rates2010, on='GEO_ID')
#get column names
cols = tracts_merged.columns.tolist()
#select only columns we want
cols = ['Geo_Name','county', 'State_y', 'Region', '2020_tract_rate', '2010_rate', '2020_state_rate', '10_20_state_difference', '2020_tract_st_diff']
tracts_merged = tracts_merged[cols]
#rename weird column name
tracts_merged = tracts_merged.rename(columns={'State_y':'State', '2020_rate':'2020_tract_rate', '2010_rate':'2010_tract_rate'})
#print df sorted largest --> smallest 2010 rate
tracts_merged.sort_values(by='2010_tract_rate', ascending=False)

Geo_Name                  county        State  \
10379  Tract 109.06, San Bernardino   San Bernardino County   California   
48849             Tract 9400, Otero            Otero County   New Mexico   
13467          Tract 9800, Arapahoe         Arapahoe County     Colorado   
8287           Tract 9800, Monterey         Monterey County   California   
6449     Tract 2653.01, Los Angeles      Los Angeles County   California   
...                             ...                     ...          ...   
83055        Tract 9402.02, Fremont          Fremont County      Wyoming   
83056        Tract 9403.01, Fremont          Fremont County      Wyoming   
83057        Tract 9403.02, Fremont          Fremont County      Wyoming   
83058           Tract 9404, Fremont          Fremont County      Wyoming   
83059           Tract 9405, Fremont          Fremont County      Wyoming   

      Region  2020_tract_rate  2010_tract_rate  2020_state_rate  \
10379   West             10.8            100.0             62.4   
48849   West             15.9            100.0             49.8   
13467   West             87.5            100.0             64.6   
8287    West              0.7            100.0             62.4   
6449    West              4.8            100.0             62.4   
...      ...              ...              ...              ...   
83055   West             19.5              NaN             54.5   
83056   West             27.8              NaN             54.5   
83057   West             36.3              NaN             54.5   
83058   West             46.5              NaN             54.5   
83059   West             41.9              NaN             54.5   

       10_20_state_difference  2020_tract_st_diff  
10379                   -10.6               -51.6  
48849                   -15.2               -33.9  
13467                    -7.4                22.9  
8287                    -10.6               -61.7  
6449                    -10.6               -57.6  
...                       ...                 ...  
83055                   -14.5               -35.0  
83056                   -14.5               -26.7  
83057                   -14.5               -18.2  
83058                   -14.5                -8.0  
83059                   -14.5               -12.6  

[84093 rows x 9 columns]

In [59]:
#how many null 2010 response values are there: 531, which is .6% of all rows, 84093
is_no_2010 = tracts_merged.isnull()
no_2010 = is_no_2010.any(axis=1)
no_2010 = tracts_merged[no_2010]
no_2010.sort_values(by="2010_tract_rate")

#due to the low percentage, these null values will be discarded
tracts_merged = tracts_merged.dropna(axis=0)
#check they'd discraded
tracts_merged.sort_values(by='2010_tract_rate')
#there are no 2010 na values, so all were dropped

Geo_Name               county          State     Region  \
77029      Tract 9663, Windsor       Windsor County        Vermont  Northeast   
48861   Tract 9408, Rio Arriba    Rio Arriba County     New Mexico       West   
6111   Tract 2074, Los Angeles   Los Angeles County     California       West   
21943     Tract 231.15, DeKalb        DeKalb County        Georgia      South   
63439        Tract 122, Centre        Centre County   Pennsylvania  Northeast   
...                        ...                  ...            ...        ...   
13467     Tract 9800, Arapahoe      Arapahoe County       Colorado       West   
24218          Tract 7.03, Ada           Ada County          Idaho       West   
73376    Tract 202.07, Hidalgo       Hidalgo County          Texas      South   
50116       Tract 9704, Fulton        Fulton County       New York  Northeast   
49471         Tract 435, Bronx         Bronx County       New York  Northeast   

       2020_tract_rate  2010_tract_rate  2020_state_rate  \
77029             20.4              0.0             54.3   
48861             22.2              0.0             49.8   
6111              25.9              0.0             62.4   
21943             10.0              0.0             57.4   
63439             17.6              0.0             64.7   
...                ...              ...              ...   
13467             87.5            100.0             64.6   
24218             77.8            100.0             64.9   
73376             36.5            100.0             56.1   
50116             19.6            100.0             56.6   
49471             61.0            100.0             56.6   

       10_20_state_difference  2020_tract_st_diff  
77029                   -14.7               -33.9  
48861                   -15.2               -27.6  
6111                    -10.6               -36.5  
21943                   -14.6               -47.4  
63439                   -12.3               -47.1  
...                       ...                 ...  
13467                    -7.4                22.9  
24218                   -11.1                12.9  
73376                   -14.9               -19.6  
50116                   -12.4               -37.0  
49471                   -12.4                 4.4  

[83562 rows x 9 columns]

In [60]:
#create column with difference in 2010 vs 2020 response rate
tracts_merged['10_20_tract_difference'] = tracts_merged['2020_tract_rate'] - tracts_merged['2010_tract_rate']
#sort df largest --> smallest 10-20 difference
tracts_merged.sort_values(by='10_20_tract_difference', ascending=False)

Geo_Name                county          State  \
23292      Tract 106.06, Muscogee       Muscogee County        Georgia   
22281            Tract 37, Fulton         Fulton County        Georgia   
65389     Tract 364, Philadelphia   Philadelphia County   Pennsylvania   
40288           Tract 9824, Wayne          Wayne County       Michigan   
21302       Tract 107.07, Chatham        Chatham County        Georgia   
...                           ...                   ...            ...   
6449   Tract 2653.01, Los Angeles    Los Angeles County     California   
53795      Tract 1595.10, Suffolk        Suffolk County       New York   
49059          Tract 4.04, Albany         Albany County       New York   
7980   Tract 9800.21, Los Angeles    Los Angeles County     California   
8287         Tract 9800, Monterey       Monterey County     California   

          Region  2020_tract_rate  2010_tract_rate  2020_state_rate  \
23292      South             98.1              0.0             57.4   
22281      South             67.3              0.0             57.4   
65389  Northeast             80.2             13.5             64.7   
40288    Midwest             83.1             16.7             67.7   
21302      South             68.1              2.4             57.4   
...          ...              ...              ...              ...   
6449        West              4.8            100.0             62.4   
53795  Northeast              2.2            100.0             56.6   
49059  Northeast              1.8            100.0             56.6   
7980        West              1.6            100.0             62.4   
8287        West              0.7            100.0             62.4   

       10_20_state_difference  2020_tract_st_diff  10_20_tract_difference  
23292                   -14.6                40.7                    98.1  
22281                   -14.6                 9.9                    67.3  
65389                   -12.3                15.5                    66.7  
40288                   -10.3                15.4                    66.4  
21302                   -14.6                10.7                    65.7  
...                       ...                 ...                     ...  
6449                    -10.6               -57.6                   -95.2  
53795                   -12.4               -54.4                   -97.8  
49059                   -12.4               -54.8                   -98.2  
7980                    -10.6               -60.8                   -98.4  
8287                    -10.6               -61.7                   -99.3  

[83562 rows x 10 columns]

## Demographic Data

In [61]:
#load both sets of demographic data, join
demo1 = pd.read_csv('https://raw.githubusercontent.com/tdiffendal/USAT/master/census-responses/demographics1_working.csv')
demo2 = pd.read_csv('https://raw.githubusercontent.com/tdiffendal/USAT/master/census-responses/demographics2_working.csv')
demo = pd.merge(demo1, demo2, on=['GISJOIN', "YEAR", "STATE", "STATEA", 
                                              'COUNTY', 'COUNTYA', 'TRACTA', 
                                              'Geo_Name', 'NAME_E'])
#view demographics table
demo

# merge to create new df with response rates and demos for all tracts
df = pd.merge(tracts_merged, demo, on='Geo_Name')

#create new column adding up pop with rent > 30% income (homelessness marker)
df['rent_30_more'] = (df['rent_30_34.9'] + df['rent_35_39.9'] 
+ df['rent_40_49.9'] + df['rent_50_over'])
# check to see if column created
pd.set_option('display.max_columns', 100)
df

Geo_Name            county         State       Region  \
0        Tract 201, Autauga    Autauga County       Alabama        South   
1        Tract 202, Autauga    Autauga County       Alabama        South   
2        Tract 203, Autauga    Autauga County       Alabama        South   
3        Tract 204, Autauga    Autauga County       Alabama        South   
4        Tract 206, Autauga    Autauga County       Alabama        South   
...                     ...               ...           ...          ...   
64408  Tract 7505.01, Yauco   Yauco Municipio   Puerto Rico  Puerto Rico   
64409  Tract 7505.02, Yauco   Yauco Municipio   Puerto Rico  Puerto Rico   
64410  Tract 7505.03, Yauco   Yauco Municipio   Puerto Rico  Puerto Rico   
64411  Tract 7506.01, Yauco   Yauco Municipio   Puerto Rico  Puerto Rico   
64412  Tract 7506.02, Yauco   Yauco Municipio   Puerto Rico  Puerto Rico   

       2020_tract_rate  2010_tract_rate  2020_state_rate  \
0                 64.3             70.6             59.2   
1                 65.6             70.1             59.2   
2                 73.8             73.6             59.2   
3                 77.8             78.4             59.2   
4                 66.2             71.9             59.2   
...                ...              ...              ...   
64408             29.9             70.4             18.9   
64409             36.2             73.4             18.9   
64410             20.9             63.4             18.9   
64411             21.9             67.1             18.9   
64412             14.6             57.8             18.9   

       10_20_state_difference  2020_tract_st_diff  10_20_tract_difference  \
0                       -12.8                 5.1                    -6.3   
1                       -12.8                 6.4                    -4.5   
2                       -12.8                14.6                     0.2   
3                       -12.8                18.6                    -0.6   
4                       -12.8                 7.0                    -5.7   
...                       ...                 ...                     ...   
64408                   -35.1                11.0                   -40.5   
64409                   -35.1                17.3                   -37.2   
64410                   -35.1                 2.0                   -42.5   
64411                   -35.1                 3.0                   -45.2   
64412                   -35.1                -4.3                   -43.2   

              GISJOIN       YEAR        STATE  STATEA           COUNTY  \
0      G0100010020100  2014-2018      Alabama       1   Autauga County   
1      G0100010020200  2014-2018      Alabama       1   Autauga County   
2      G0100010020300  2014-2018      Alabama       1   Autauga County   
3      G0100010020400  2014-2018      Alabama       1   Autauga County   
4      G0100010020600  2014-2018      Alabama       1   Autauga County   
...               ...        ...          ...     ...              ...   
64408  G7201530750501  2014-2018  Puerto Rico      72  Yauco Municipio   
64409  G7201530750502  2014-2018  Puerto Rico      72  Yauco Municipio   
64410  G7201530750503  2014-2018  Puerto Rico      72  Yauco Municipio   
64411  G7201530750601  2014-2018  Puerto Rico      72  Yauco Municipio   
64412  G7201530750602  2014-2018  Puerto Rico      72  Yauco Municipio   

       COUNTYA  TRACTA                            NAME_E  total_population  \
0            1   20100        Tract 201, Autauga Alabama              1923   
1            1   20200        Tract 202, Autauga Alabama              2028   
2            1   20300        Tract 203, Autauga Alabama              3476   
3            1   20400        Tract 204, Autauga Alabama              3831   
4            1   20600        Tract 206, Autauga Alabama              3705   
...        ...     ...                               ...               ...   
64408      153  750501  T

In [62]:
#df.std()Returns the standard deviation of each column
#df.corr()Returns the correlation between columns in a data frame

#In order to get # of null/missing values for each column, run 
pd.set_option('display.max_rows', 100)
pd.isnull(df).sum()

# make a dataframe of all rows with na value
df1 = df[df.isna().any(axis=1)]
df1
# how many nas in each state
hm = df1['State'].value_counts()
hm = pd.DataFrame(hm)
hm = hm.reset_index().rename(columns={'index':'state', 'State':'count_na'})
hm
#compare to total 
hmm = df['State'].value_counts()
hmm = pd.DataFrame(hmm)
hmm = hmm.reset_index().rename(columns={'index':'state', 'State':'count'})
hmm

#see what percent of state values of na
#Puerto Rico will lose the greatest % if these are dropped
hmmm = pd.merge(hm, hmm, on=['state'])
hmmm['na_percent'] = (hmmm['count_na']*100) / hmmm['count']
hmmm

#how many total nas? --> 1171
sum(hmmm['count_na'])

#nas account for what total % of all rows --> 1.82
(sum(hmmm['count_na'])*100) / sum(hmmm['count'])

#Less than 2% of total, so for now will drop those rows
# before df had 64413 rows
df = df.dropna()    #now has 63242 rows, 1171 difference (total # nas)

In [63]:
#Df column names for reference
#dflist = df.columns.tolist()
#dflist

#print df with no nas
#df.to_csv("all_rates_demos_merged.csv")

#create df with just puerto rico
is_pr = df['Region'] == 'Puerto Rico'
pr = df[is_pr]
pr

#df without puerto rico
regdf = df[df['Region'] != 'Puerto Rico']
regdf

Geo_Name            county     State Region  \
0        Tract 201, Autauga    Autauga County   Alabama  South   
1        Tract 202, Autauga    Autauga County   Alabama  South   
2        Tract 203, Autauga    Autauga County   Alabama  South   
3        Tract 204, Autauga    Autauga County   Alabama  South   
4        Tract 206, Autauga    Autauga County   Alabama  South   
...                     ...               ...       ...    ...   
63583     Tract 2, Washakie   Washakie County   Wyoming   West   
63584  Tract 3.01, Washakie   Washakie County   Wyoming   West   
63585  Tract 3.02, Washakie   Washakie County   Wyoming   West   
63586    Tract 9511, Weston     Weston County   Wyoming   West   
63587    Tract 9513, Weston     Weston County   Wyoming   West   

       2020_tract_rate  2010_tract_rate  2020_state_rate  \
0                 64.3             70.6             59.2   
1                 65.6             70.1             59.2   
2                 73.8             73.6             59.2   
3                 77.8             78.4             59.2   
4                 66.2             71.9             59.2   
...                ...              ...              ...   
63583             54.0             72.3             54.5   
63584             60.6             73.3             54.5   
63585             66.4             80.3             54.5   
63586             42.1             68.3             54.5   
63587             51.8             68.5             54.5   

       10_20_state_difference  2020_tract_st_diff  10_20_tract_difference  \
0                       -12.8                 5.1                    -6.3   
1                       -12.8                 6.4                    -4.5   
2                       -12.8                14.6                     0.2   
3                       -12.8                18.6                    -0.6   
4                       -12.8                 7.0                    -5.7   
...                       ...                 ...                     ...   
63583                   -14.5                -0.5                   -18.3   
63584                   -14.5                 6.1                   -12.7   
63585                   -14.5                11.9                   -13.9   
63586                   -14.5               -12.4                   -26.2   
63587                   -14.5                -2.7                   -16.7   

              GISJOIN       YEAR    STATE  STATEA           COUNTY  COUNTYA  \
0      G0100010020100  2014-2018  Alabama       1   Autauga County        1   
1      G0100010020200  2014-2018  Alabama       1   Autauga County        1   
2      G0100010020300  2014-2018  Alabama       1   Autauga County        1   
3      G0100010020400  2014-2018  Alabama       1   Autauga County        1   
4      G0100010020600  2014-2018  Alabama       1   Autauga County        1   
...               ...        ...      ...     ...              ...      ...   
63583  G5600430000200  2014-2018  Wyoming      56  Washakie County       43   
63584  G5600430000301  2014-2018  Wyoming      56  Washakie County       43   
63585  G5600430000302  2014-2018  Wyoming      56  Washakie County       43   
63586  G5600450951100  2014-2018  Wyoming      56    Weston County       45   
63587  G5600450951300  2014-2018  Wyoming      56    Weston County       45   

       TRACTA                        NAME_E  total_population  \
0       20100    Tract 201, Autauga Alabama              1923   
1       20200    Tract 202, Autauga Alabama              2028   
2       20300    Tract 203, Autauga Alabama              3476   
3       20400    Tract 204, Autauga Alabama              3831   
4       20600    Tract 206, Autauga Alabama              3705   
...       ...                           ...               ...   
63583     200     Tract 2, Washakie Wyoming              3098   
63584     301  Tract 3.01, Washakie Wyoming              2447   
63585     302  Tract 3.02, Washakie Wyoming   

## Stats

In [64]:
# Average difference between current state rates and 2010 state rates as of 6/15/20
df.mean(axis=0)['10_20_state_difference']

KeyboardInterrupt: ignored

In [65]:
# average difference by region
states_merged.groupby('Region').mean().sort_values(by='10_20_state_difference', ascending=False)

2020_state_rate  2010_state_rate  2000_state_rate  \
Region                                                           
Midwest            65.723077        77.307692        79.230769   
West               59.207692        70.923077        72.538462   
Northeast          59.550000        72.375000        72.875000   
South              58.629412        72.176471        70.647059   
Puerto Rico        18.900000        54.000000        54.000000   

             10_20_state_difference  
Region                               
Midwest                  -11.584615  
West                     -11.715385  
Northeast                -12.825000  
South                    -13.547059  
Puerto Rico              -35.100000

#### State stats without Puerto Rico

In [ ]:
no_pr = states[states.State != 'Puerto Rico']

# average non internet response rate
no_pr.mean(axis=0)['state_not_int']

In [ ]:
# average internet response
no_pr.mean(axis=0)['state_internet']

In [ ]:
# average overall response rate
no_pr.mean(axis=0)['2020_state_rate']

In [ ]:
# average region response rate (so south now excludes Puerto Rico)
no_pr.groupby('Region').mean().sort_values(by='2020_state_rate', ascending=False)

In [ ]:
# highest non-internet response rate (not_int)
states.sort_values(by='state_not_int', ascending=False)

In [ ]:
states.mean(axis=0)['state_not_int']

In [ ]:
states.mean(axis=0)['state_internet']

In [ ]:
states.mean(axis=0)['2020_state_rate']

In [ ]:
# average by region
# NOTE as tribal tracts are not assigned to a state they do not have a corresponding region and thus are not counted in the regional calculations
states.groupby('Region').mean().sort_values(by='2020_state_rate', ascending=False)

In [ ]:
# average difference as of 6/15/20
#this can take a while to run so is commented out unless needed
#tracts_merged.mean(axis=0)['10_20_tract_difference']

In [ ]:
# average difference by region
tracts_merged.groupby('Region').mean().sort_values(by='10_20_tract_difference', ascending=False)

In [ ]:
#tract average differences vs state rates
tracts_merged.groupby('State').mean().sort_values(by='2020_state_rate', ascending=False)

#### Tribal tracts

In [ ]:
# create df with response rates in tribal tracts
tribal = tracts[tracts['Geo_Type'].str.contains("Tribal")]
tribal.sort_values(by='2020_tract_rate', ascending=False)

In [ ]:
### tribal areas and tracts stats

#mean non internet response
tribal.mean(axis=0)['not_int']

In [ ]:
# mean internet response rate
tribal.mean(axis=0)['internet']

In [ ]:
# mean overall response rate
tribal.mean(axis=0)['2020_tract_rate']

#### Tracts with 0 overall response rate

In [ ]:
## Tracts with 0 cumulative response rate: 28
is_zero = tracts['2020_tract_rate'] == 0.0
zeros = tracts[is_zero]
zeros.sort_values(by='State')

## Tracts with 0 cumulative response rate: 28

#make dataframe of states with # tracts with 0%, number total tracts, and what % of total tracts are 0
temp = pd.DataFrame(zeros['State'].value_counts())
temp2 = pd.DataFrame(tracts['State'].value_counts())
temp3 = pd.merge(temp, temp2, right_index=True, left_index=True)
#rename the columns
temp3 = temp3.rename(columns={"State_x": "0_tracts", "State_y" : "total_tracts"})
#compute percentage
temp3['0_percent'] = temp3['0_tracts'] * 100 / temp3['total_tracts']
temp3

## Regressions

### 2020 regressions

In [ ]:
### 2020 Multi-regression

import statsmodels.api as sm

#put all variables for predicting 2020 rates in dataframe
variables20 = regdf[['2010_tract_rate','total_population', 
 'white_alone', 'black_alone', 'amerindian_alone', 'asian_alone', 
 'pacific_islander_alone',
 'other_alone', 'two_or_more', 'two_more_including_other',
 'two_more_excluding_other', 'total_education', 'no_school',
 'some_school', 'diploma', 'ged', 'some_college', 'associate',
 'bachelor', 'master', 'prof_school', 'doctorate', 'language_total',
 'lang_english_only', 'lang_spanish', 'lang_spanish_limited_english', 
 'lang_other_indo_euro', 'lang_asian_pacific_island',
 'lang_other', 'income_poverty_ratio', 'income_poverty_under_half',
 'income_poverty_half_.99', 'income_povery_1_1.24',
 'income_poverty_1.25_1.49', 'income_poverty_1.5_1.84',
 'income_poverty_1.85_1.99', 'income_poverty_2_over',
 'median_household_income', 'per_capita_income',
 'employment_total', 'labor_force', 'civilian_labor_force',
 'civilian_employed', 'civilian_unemployed', 'armed_forces',
 'not_labor_force', 'total_houses', 'occupied_houses',
 'vacant_houses', 'total_occupied_houses', 'owner_occupied',
 'renter_occupied', 'median_gross_rent', 'rent_to_income',
 'rent_less_10', 'rent_10_14.9', 'rent_15_19.9', 'rent_20_24.9',
 'rent_25_29.9', 'rent_30_34.9', 'rent_35_39.9', 'rent_40_49.9',
 'rent_50_over', 'rent_30_more', 'rent_not_computed', 'total_computer_status',
 'has_computer', 'dial_up_computer', 'broadband_computer',
 'no_internet_computer', 'no_computer', 'us_pop',
 'us_born', 'us_territory_born', 'us_born_abroad',
 'us_naturalization', 'not_us_citizen']]

#what we want to predict - 2020 response rates - in dataframe
target20 = regdf[["2020_tract_rate"]]

#build model and print summary
model20 = sm.OLS(target20, variables20).fit()
model20.summary()

In [ ]:
### 2020 linear regressions for each variable

from sklearn import linear_model

#create list of variable names
cols = variables20.columns.tolist()
#build linear model
regr = linear_model.LinearRegression()
#create empty list to store loop results
rows = []
#loop through each variable
for i in cols:
    #fit linear model to variable
    regr.fit(variables20[[i]], target20)
    #save model variable name, intercept, coef and r^2 to list
    rows.append([i, regr.intercept_, regr.coef_, regr.score(variables20[[i]], target20)])

#turn list into df with these column names
linears20 = pd.DataFrame(rows, columns=['variable', 'intercept', 'coefficient', 'r-squared'])

#remove square brackets lol
linears20['coefficient'] = linears20['coefficient'].str.get(0)
linears20['coefficient'] = linears20['coefficient'].str.get(0)
linears20['intercept'] = linears20['intercept'].str.get(0)

#print df sorted coefs largest --> smallest
linears20.sort_values(by='coefficient', ascending=False)

### Normalize inputs

In [ ]:
### 2020 Multi regression with normalized variables

#normalize variable values
norm_variables20 = (variables20 - variables20.min()) / (variables20.max() - variables20.min())

#build normalized multi-regress model and print summary
norm_model20 = sm.OLS(target20, norm_variables20).fit()
norm_model20.summary()

In [ ]:
#2020 normalized linear regressions for each variable

# linear regression for each variable 'i'
cols = norm_variables20.columns.tolist()
# create model
norm_regr = linear_model.LinearRegression()
#empty list for loop results
rows = []
#loop through each variable
for i in cols:
    #fit model to each variable
    norm_regr.fit(norm_variables20[[i]], target20)
    #add model results to list
    rows.append([i, norm_regr.intercept_, norm_regr.coef_, 
                 norm_regr.score(norm_variables20[[i]], target20)])

#turn list into dataframe
norm_linears20 = pd.DataFrame(rows, columns=['variable', 'intercept', 'coefficient', 'r-squared'])

#remove square brackets
norm_linears20['coefficient'] = norm_linears20['coefficient'].str.get(0)
norm_linears20['coefficient'] = norm_linears20['coefficient'].str.get(0)
norm_linears20['intercept'] = norm_linears20['intercept'].str.get(0)

#print df ordered by largest to smallest coef
norm_linears20.sort_values(by='coefficient', ascending=False)

### 2010 Regressions

In [ ]:
### 2010 multi regression

#put all variables for predicting 2010 rates in dataframe
variables10 = regdf[['total_population', 
 'white_alone', 'black_alone', 'amerindian_alone', 'asian_alone', 
 'pacific_islander_alone',
 'other_alone', 'two_or_more', 'two_more_including_other',
 'two_more_excluding_other', 'total_education', 'no_school',
 'some_school', 'diploma', 'ged', 'some_college', 'associate',
 'bachelor', 'master', 'prof_school', 'doctorate', 'language_total',
 'lang_english_only', 'lang_spanish', 'lang_spanish_limited_english', 
 'lang_other_indo_euro', 'lang_asian_pacific_island',
 'lang_other', 'income_poverty_ratio', 'income_poverty_under_half',
 'income_poverty_half_.99', 'income_povery_1_1.24',
 'income_poverty_1.25_1.49', 'income_poverty_1.5_1.84',
 'income_poverty_1.85_1.99', 'income_poverty_2_over',
 'median_household_income', 'per_capita_income',
 'employment_total', 'labor_force', 'civilian_labor_force',
 'civilian_employed', 'civilian_unemployed', 'armed_forces',
 'not_labor_force', 'total_houses', 'occupied_houses',
 'vacant_houses', 'total_occupied_houses', 'owner_occupied',
 'renter_occupied', 'median_gross_rent', 'rent_to_income',
 'rent_less_10', 'rent_10_14.9', 'rent_15_19.9', 'rent_20_24.9',
 'rent_25_29.9', 'rent_30_34.9', 'rent_35_39.9', 'rent_40_49.9',
 'rent_50_over', 'rent_30_more', 'rent_not_computed', 'total_computer_status',
 'has_computer', 'dial_up_computer', 'broadband_computer',
 'no_internet_computer', 'no_computer', 'us_pop',
 'us_born', 'us_territory_born', 'us_born_abroad',
 'us_naturalization', 'not_us_citizen']]

#what we want to predict - 2010 response rates - in separate dataframe
target10 = regdf[["2010_tract_rate"]]

#create model and print summary table
model10 = sm.OLS(target10, variables10).fit()
model10.summary()

In [ ]:
### 2010 linear regression for each variable

#list of variable names
cols = variables10.columns.tolist()
#build multi-reg model
regr = linear_model.LinearRegression()
#create empty list for loop results
rows = []

#loop through variables
for i in cols:
    #fit a model to the current variable
    regr.fit(variables10[[i]], target10)
    #save the model's resulting variable name, intercept, coef, and r^2
    rows.append([i, regr.intercept_, regr.coef_,
                regr.score(variables10[[i]], target10)])

#turn list into data frame
linears10 = pd.DataFrame(rows, columns=['variable', 'intercept', 'coefficient', 'r-squared'])

#remove square brackets
linears10['coefficient'] = linears10['coefficient'].str.get(0)
linears10['coefficient'] = linears10['coefficient'].str.get(0)
linears10['intercept'] = linears10['intercept'].str.get(0)

#print data frame ordered coefficient largest --> smallest
linears10.sort_values(by='coefficient', ascending=False)

### Normalized 2010 Regressions

In [ ]:
### 2010 normalized multi-regression

#normalize the variables
norm_variables10 = (variables10 - variables10.min()) / (variables10.max() - variables10.min())

#build normalized model and print summary
norm_model10 = sm.OLS(target10, norm_variables10).fit()
norm_model10.summary()

In [ ]:
###2010 normalized linear regressions for each variable

#create list of variable names
cols = norm_variables10.columns.tolist()
#build the model
norm_regr = linear_model.LinearRegression()
#create empty list for model results
rows = []
#cycle through variables
for i in cols:
    #do the linear regression on the current variable
    norm_regr.fit(norm_variables10[[i]], target10)
    #add the corresponding variable name, intercept, coefficient and r-squared to the list
    rows.append([i, norm_regr.intercept_, norm_regr.coef_,
                norm_regr.score(norm_variables10[[i]], target10)])

#turn list into data frame with these column names
norm_linears10 = pd.DataFrame(rows, columns=['variable', 'intercept', 'coefficient', 'r-squared'])

#remove square brackets
norm_linears10['coefficient'] = norm_linears10['coefficient'].str.get(0)
norm_linears10['coefficient'] = norm_linears10['coefficient'].str.get(0)
norm_linears10['intercept'] = norm_linears10['intercept'].str.get(0)

#print df sorted by coefficient
norm_linears10.sort_values(by='coefficient', ascending=False)

## Edited Regressions
### Run regressions with fewer variables

In [ ]:
### 2020 edited multi-regressions

#put all variables for predicting 2020 rates in dataframe
variables_ed = regdf[['2010_tract_rate','total_population', 
 'white_alone', 'black_alone', 'amerindian_alone', 'asian_alone', 
 'pacific_islander_alone', 'no_school',
 'some_school', 'diploma', 'ged', 'some_college', 'associate',
 'bachelor', 'master',
 'lang_english_only', 'lang_spanish', 'lang_spanish_limited_english', 
 'lang_other', 'income_poverty_ratio', 'per_capita_income',
 'civilian_employed', 'civilian_unemployed',
 'not_labor_force', 'total_houses', 'occupied_houses',
 'vacant_houses', 'owner_occupied',
 'renter_occupied', 'median_gross_rent', 'rent_to_income','rent_30_more',
 'has_computer', 'dial_up_computer', 'broadband_computer',
 'no_internet_computer', 'no_computer',
 'us_born', 'us_naturalization', 'not_us_citizen']]

#what we want to predict - 2020 response rates - in dataframe
target_ed = regdf[["2020_tract_rate"]]

#build and fit the multi-regression model
model_ed = sm.OLS(target_ed, variables_ed).fit()
#print out the model summary table
model_ed.summary()

In [ ]:
### 2020 edited linear regressions

#create list of variable names
cols = variables_ed.columns.tolist()
#build the model
regr = linear_model.LinearRegression()
#create empty list to append results
rows = []

#loop through variables
for i in cols:
    #fit the model
    regr.fit(variables_ed[[i]], target_ed)
    #put model variable name, intercept, coef and r^2 in list
    rows.append([i, regr.intercept_, regr.coef_,
                regr.score(variables_ed[[i]], target_ed)])

#turn list into data frame with these column names
linears_ed = pd.DataFrame(rows, columns=['variable', 'intercept', 'coefficient', 'r-squared'])

#remove square brackets
linears_ed['coefficient'] = linears_ed['coefficient'].str.get(0)
linears_ed['coefficient'] = linears_ed['coefficient'].str.get(0)
linears_ed['intercept'] = linears_ed['intercept'].str.get(0)

#print df sorted by coefficient largest --> smallest
linears_ed.sort_values(by='coefficient', ascending=False)

In [ ]:
### normalized 2020 edited variables multi regression

#normalize the variables
norm_variables_ed = (variables_ed - variables_ed.min()) / (variables_ed.max() - variables_ed.min())

#build normalized model and print summary
norm_model_ed = sm.OLS(target_ed, norm_variables_ed).fit()
norm_model_ed.summary()